modules/core/src/main/java/org/apache/ignite/spi/encryption/EncryptionSpi.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.spi.encryption;
+
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.spi.IgniteSpi;
+
+/**
+ * SPI provides encryption features for an Ignite.
+ */
+public interface EncryptionSpi extends IgniteSpi {
+    /**
+     * Returns master key digest.
+     * Should always return same digest for a same key.
+     * Digest used for a configuration consistency check.
+     *
+     * @return Master key digest.
+     */
+    byte[] masterKeyDigest();
+
+    /**
+     * Creates new key for an encryption/decryption of cache persistent data: pages, WAL records.
+     *
+     * @return Newly created encryption key.
+     * @throws IgniteException If key creation failed.
+     */
+    Serializable create() throws IgniteException;
+
+    /**
+     * Encrypts data.
+     * 
+     * @param data Data to encrypt.
+     * @param key Encryption key.
+     * @param res Destination buffer.
+     */
+    void encrypt(ByteBuffer data, Serializable key, ByteBuffer res);
```

Comment:
why encrypt takes as parameter ByteBuffers, but decrypt takes byte[]? Can we add uniformity and implement symmetric method pairs with same parameter types?

Reply:
Because of other internal parts of Ignite.
Do you think we should try to make this API convenient for some kind of "end-user"?

Reply:
I think API should be uniform. It is not an API for an end user, but it is still API for an advanced user. 

If we have different methods but taking a different type of parameters an implementor can be confused by such differences. It will raise a question, why this method takes byte[] but this method takes ByteBuffer.